hotstuff-rs 0.4 new module organisation and types

karolinagrzeszkiewicz opened the pull request:
This PR proposes a new code organisation that is expected to enable a smooth integration of the new components we intend to introduce with hotstuff-rs 0.4. In particular, this enables a `Pacemaker` protocol, with associated networking functionality, to replace what used to be a `Pacemaker` trait with two associated methods.

**The scope of this PR** is as follows:
1. New, more granular, modular structure,
2. New `HotStuff`, `Pacemaker`, `BlockSyncClient`, `BlockSyncServer` types,
3. New message type hierarchy,
4. New generics to generalize some functionalities shared across the different protocols, e.g., `Collector` or `SignedMessage` traits,
5. New `Algorithm` struct which provides handles to its sub-protocols, and its `execute` method, which defines how the sub-protocols interact.
6. A fix for #30.

**Note:**
1. The `block_sync::client` and `block_sync::servers` type and method signature is very tentative, it will likely change after we finalize the implementation details.
2. This does not include the changes to the type signature that may be required to adapt dynamic validator sets (consensus and speculation). This is to enable me to test the new `Pacemaker` against the current HotStuff implementation before I update the implementation.
3. This does not include updates to state (for this PR the module is exactly the same as in 0.3). I plan to make relevant updates incrementally, i.e., the `Pacemaker`-related updates when implementing the `Pacemaker`, and then Dynamic-Validator-Sets-related updates when implementing them.
4. Documentation polishing and event handlers updates are also left for later stages of implementation.

karolinagrzeszkiewicz commented:
Some changes that I have in mind as of this moment:
1. Rename `Algorithm` to `ConsensusEngine`, `ConsensusMaker`, or `ProtocolExecutor`, 
2. Use `Rc<T>` for storing a pointer to block_tree as a field of `HotStuff`, `Pacemaker`, and `SyncClient`, instead of passing a mutable reference to the block tree every time we call their methods.

lyulka commented:
I skimmed through the PR a bit on my phone but will have to continue tomorrow. Two points I want to look at:
- There are some interesting traits in the PR like Collector, SignedMessage, and Certificate. I wonder what sort of generality these traits enable in the code.
- I wonder whether we can make the creation of Nudges a bit more type-safe by limiting the kinds of QCs that a Nudge can contain at a type level, instead of panicking in the constructor like we currently do (behavior which this PR doesn’t change).

karolinagrzeszkiewicz commented:
> I skimmed through the PR a bit on my phone but will have to continue tomorrow. Two points I want to look at (will definitely look at other things, too):
> 
>     * There are some interesting traits in the PR like Collector, SignedMessage, and Certificate. Each of these link together a couple/few different types in HotStuff-rs based on their common functionalities. This seems nice, but I wonder concretely what sort of generality/succinctness these traits enable in the code.
> 
>     * I wonder whether we can make the creation of Nudges a bit more type-safe by limiting the kinds of QCs that a Nudge can contain at a type level, instead of panicking in the constructor like we currently do (behavior which this PR doesn’t change).

#### Traits

The traits introduce more generality and extensibility, but they don't enable succinctness as much as I'd want to. It mostly serves to generalize types, e.g, that QC and TC are really instances of the same thing. More generally they point to the fact that HotStuff and Pacemaker operate based on the some shared principles of consensus, such as having certificates and signed messages serve as evidence. They also enforce a shared signature among instances. It comes with some provided methods too, e.g., `SignedMessage::is_correct` and `Certificate::quorum`. 

Ideally I'd like the simple, small methods to be required, e.g., `SignedMessage::signature_bytes`, and all "big" methods to be provided given the required methods. Concretely, I'd like `Collector::collect` and `Certificate::is_correct` to be provided. The reason why I haven't done it is that it requires adding more required methods like `Certificate::value_bytes` and `Certificate::is_genesis` (always false for a TC), but now that I think about it adding those methods shouldn't big a big deal. What do you think?

#### Type safety and nudges

Yes, I've had this in mind for the past few weeks, but I didn't include it since I thought we postponed the decision about type wrapping for later and I am also not sure about the best way to go about this... Ideally I would enforce everything at the type-level, especially since a lot of bugs I have seen so far in hotstuff-rs were due to type aliasing or confusion regarding the meaning of a type.

What I have in mind is to have structs `NudgeJustify {qc: QuorumCertificate}` and `BlockJustify {qc: QuorumCertificate}` (possibly also `AdvanceViewJustify` for consistency), and implement `TryFrom`/ `TryInto`. However, wrapping may not be enough, since this way a `NudgeJustify `can still have a `qc` with a `Generic` phase, but it will fail `QuorumCertificate::try_into` (and it seems wrong to return `Err `here, since it clearly contains a QC). 

Perhaps more intuitive solution would be to have a `QuorumCertificate` enum or trait, which branches out into a `NudgeJustify` and `BlockJustify`, which are like the current `QuorumCertificate` struct but the former has `NudgeJustifyPhase` and the latter has `BlockJustifyPhase`.

Or more elegantly, `QuorumCertificate<P: PhaseType>` where `enum PhaseType {BlockJustifyPhase(BlockJustifyPhase), NudgeJustifyPhase(NudgeJustifyPhase)}`.

Thoughts?

karolinagrzeszkiewicz commented:
**Note: this comment is a work in progress, I am about to add some things.**

Let me respond here, just to keep a record, but we should also discuss this during a meeting.

## About the `SignedMessage`, `Certificate` and `Collector` traits

Point taken, I don't think the `Collector` is useful in its current form. Indeed, `NewViewCollector` does not fit the definition because the `NewView` messages are not signed.

## Blanket implementation for `Collector::collect`

As for introducing blanket implementations for `Collector` or introducing a `Collector` struct generic over `Vote`, I am realizing that it actually involves adding significant complexity, i.e., more required methods. The question is hwther it is worth it. It might make more sense to remove the trait as you originally suggested, and have a `VoteCollector`/`BlockVoteCollector,` `TimeoutVoteCollector` and `NewViewCollector`. 

The `VoteCollector `struct must have some "signature storage" as a field. However, this varies between the `BlockVoteCollector` and the `TimeoutVoteCollector`. The former stores possibly multiple signature sets for different blocks and phases in a `HashMap<(CryptoHash, Phase), (SignatureSet, TotalPower)>`, and the latter stores only one signature set in `(SignatureSet, TotalPower)`. Perhaps `SignatureCache` can be an associated type of `Vote`, then we could have a different type depending on the collector, but it seems that this shouldn't really belong to the specification of a vote... So I imagine the signature would have to be something like this:

```rust
trait Vote {
    type Certificate: Certificate;
    type SignatureCache: SignatureCache;
   
    // Omitted: methods common to both HotStuffVote and TimeoutVote.
    fn insert_to_signature_cache(self,  &mut signature_cache: SignatureCache) -> (SignatureSet, TotalPower);
}

trait Certificate {

   // other methods, provided earlier in my implementation

    fn contains_certificate(signature_set: (SignatureSet, TotalPower)) -> Option<Self> {
            // 1. Check if there is a quorum
           // 2. If yes construct a Certificate
    };

struct VoteCollector<V: Vote> {
     view: ViewNumber,
     chain_id: ChainID,
     signature_cache: SignatureCache,
}

impl<V: Vote> VoteCollector<V> {

    fn collect(signer: &VerifyingKey, vote: V) -> V::Certificate {
    // 1. Check chain id and view
  
    // 2. Insert the signature into the signatures cache with insert_to_signature_cache, which should return the updated signature set

    // 3. Check if a certificate was collected with contains_certificate on the signature set returned in step 2, and if yes return it

    }

    // Omitted: constructor method for VoteCollector.
}
```

note: 
- require the SignedMessage trait, because there are signed messages i.e. advertise block, than are not votes.

karolinagrzeszkiewicz commented:
@lyulka I implemented the newtype pattern for types in `types/basic.` The primary motivation behind this implementation is that I would want to have a fixed API that restricts what can be done to values of these types (as much as possible, only to the functionalities we actually need), and avoids the downside of having to access these values through a more verbose syntax. Hence, this API is defined primarily through methods, not traits. I also followed the suggestion from [this post](https://users.rust-lang.org/t/understanding-the-perils-of-deref/47958/6) regarding implementing traits such as `Deref` for newtype pattern.

The API might be slightly extended as I continue implementing hotstuff-rs 0.4, but I don't expect significant changes.

Let me know what you think about the API? And also about the visibility of these methods?

lyulka commented:
## There should be functions to access the inner values of every newtype in `types::basic`

@karolinagrzeszkiewicz

karolinagrzeszkiewicz commented:
@lyulka point taken, I added methods for accessing the inner value and introduced consistent naming for them i.e., `get_<type>`.

I also removed the unnecessary imports in `types`. As for documentation, I intend to expand the documentation for a particular component as I implement it, so e.g., `Pacemaker` documentation will come with the next milestone, but for this milestone the documentation for `types` should be exhaustive. So let me know if you have any other feedback regarding documentation, otherwise I can also polish it as a final milestone.

I also added two methods for `QuorumCertificate`: `is_nudge_justify` and `is_block_justify`.

Let me know if this is ready to merge?

lyulka commented:
Okay, looks good to me. I'll merge now so that you can start implementing Milestone 2, but I just want to note that the Rust API Guidelines [state](https://rust-lang.github.io/api-guidelines/naming.html#getter-names-follow-rust-convention-c-getter) that "With a few exceptions, the `get_` prefix is not used for getters in Rust code."

I don't think we should take this prescription by the API guidelines as a hard rule, and I can see there being benefits (consistency, auto-code completion) to using the `get_` prefix in the case of this PR, but you may want to think about this some more and maybe make a decision about the naming of the accessors in the PR for the next milestone.

karolinagrzeszkiewicz commented:
@lyulka I didn't know about this... I had a hard time deciding what naming convention to follow, it seemed strange to access the inner type with a method name after a type like `.vec()` or `.int()`, though `.bytes()` make sense. `get_<type>` introduced some uniformity, also making the names informative, unlike `.get()` (which captures the fact that we are just **getting** the inner type, not doing any computation). I'd be reluctant to go for `as_ `or `into_` since there is no conversion happening (though this is subtle, in some sense it is a conversion). I'll think about it, what do you think would work best?
